[RDY] Data structure improvements

tarruda opened the pull request:
Some changes taken from a private branch where I refactor most libuv code to run in a background IO thread(ref #2371). Pushing as a separate PR to simplify reviewing.

tarruda committed:
test: Ensure proper initialization in unit/helpers.lua

Remove helpers.vim_init and simply perform the required initialization in
helpers.lua.

ZyX-I commented:
About rbuffer: can you consider implementing it on top of [ringbuf.h](https://github.com/ZyX-I/neovim/blob/shada/src/nvim/lib/ringbuf.h) from shada branch? I have a generic (though currently untested) ring buffer implementation there that uses macros and it is currently used for holding an array of structures (so I can’t reuse yours which holds pointers).

Though it looks like I am missing something regarding *what* is a ring buffer: `(size_t)(buf->end_ptr - buf->start_ptr)` cannot possibly be *ring* buffer size because it is OK for ring buffer to have start after the end. E.g. from [wiki](http://en.wikipedia.org/wiki/Circular_buffer):

> This image shows a full buffer with two elements having been overwritten:
> 
> ![](http://upload.wikimedia.org/wikipedia/commons/thumb/0/05/Circular_buffer_-_6789AB5_with_pointers.svg/500px-Circular_buffer_-_6789AB5_with_pointers.svg.png)

`(end - start)` here will return `-1`.

tarruda commented:
> About rbuffer: can you consider implementing it on top of ringbuf.h from shada branch? I have a generic (though currently untested) ring buffer implementation there that uses macros and it is currently used for holding an array of structures (so I can’t reuse yours which holds pointers).

I'm gonna give it a shot

tarruda commented:
> Though it looks like I am missing something regarding what is a ring buffer: (size_t)(buf->end_ptr - buf->start_ptr) cannot possibly be ring buffer size because it is OK for ring buffer to have start after the end. E.g. from wiki:

In this implementation, `end_ptr` is a pointer to the end of the array. The last element is represented by `write_ptr`("end" in your diagram)

ZyX-I commented:
@tarruda I have just pushed some fixes to my implementation. Manual test run OK after that.

tarruda commented:
@ZyX-I I'd like to keep the possibility of passing the ring buffer pointer to functions that write or read directly to memory locations, which is why I have implemented the `RBUFFER_WHILE_NOT_{FULL,EMPTY}` macros.

If the library I'm interfacing with exposes it's own internal buffer, I can just use `rbuffer_read`/`rbuffer_write` which is the case for msgpack(see channel.c/`parse_msgpack`)

But if the library expects a pointer(libvterm/`vterm_input_write` for example), I'd have to copy the memory to a temporary buffer and then pass it to the library which is a bit inefficient. With `terminal_receive`/`vterm_input_write` I usee `RBUFFER_WHILE_NOT_EMPTY`, which will feed the data by passing the pointer directly(possibly in 2 passes since it may be necessary to wrap around the buffer). 

Could you  implement something similar to `RBUFFER_WHILE_NOT_{FULL,EMPTY}` on your generic ring buffer?

ZyX-I commented:
@tarruda I guess here are the differences (yours → mine):

rbuffer_capacity() → None (using rb->buf_end - rb->buf directly when needed).
rbuffer_space() → None
rb->length → rb_length()
rbuffer_free() → rb_dealloc()
rbuffer_write_ptr(), rbuffer_read_ptr() → None (not generic)
rbuffer_produced() → None (I assume that pushing to ringbuffer is done only via rb_push or rb_insert)
rbuffer_consumed() → None (it is not assumed that reading consumes ring buffer)
rbuffer_read() → None
rbuffer_write() → rb_push() events
None → rb_free() (simple characters need not be freed)
None → rb_insert() (you do not need to insert into the middle of the ring buffer)
None → RINGBUF_ITER_BACK (you do not need to process ring buffer from the end to the start)
None → rb_remove() (you do not need to remove random elements)

Very different ring buffer implementations for different purposes.

tarruda commented:
Just found an interesting ring buffer implementation: https://github.com/michaeltyson/TPCircularBuffer

This ones uses a virtual memory technique that avoids the need for wrapping logic, I wonder if we can do something similar without introducing compatibility issues?

ZyX-I commented:
With that big amount of differences it will be easier to keep your buffer implementation, but rename it to “string ring buffer” (because mine is generic). I can add callbacks, but your variant is better suited for your purposes.

tarruda commented:
> With that big amount of differences it will be easier to keep your buffer implementation, but rename it to “string ring buffer” (because mine is generic). I can add callbacks, but your variant is better suited for your purposes.

:+1:

But later I might spend some time refactoring both into a single generic data structure. If you think about it, `rb_insert`/`rb_remove` are simply special cases of writing/reading from the buffer(could be implemented on top of `rbuffer_{read,write}` passing 1 as the size).

ZyX-I commented:
@tarruda Nope. My rb_insert/rb_remove support inserting at/removing from any index.

By the way, what is the best variant for writing tests for it? For “manual” tests I used a standalone C file with a single nvim include (ringbuf.h). Main ideas that come to mind: use this file and base test on io.open and (I guess, better): create a helper C file which will have wrapper for all tested functions (they are static inline) and use regular unit testing infrastructure, but where this is to be located?

oni-link commented on src/nvim/lib/rbuffer.h:
```diff
@@ -0,0 +1,222 @@
+    size_t copy_count = MIN(src_size, wcnt);
+    memcpy(wptr, src, copy_count);
+    rbuffer_produced(buf, copy_count);
+    src += copy_count;
```
Early return with 
```c
if (src_size == copy_count) {
   return size;
}
```
?

tarruda replied:
:+1:

tarruda commented:
> By the way, what is the best variant for writing tests for it? For “manual” tests I used a standalone C file with a single nvim include (ringbuf.h). Main ideas that come to mind: use this file and base test on io.open and (I guess, better): create a helper C file which will have wrapper for all tested functions (they are static inline) and use regular unit testing infrastructure, but where this is to be located?

I've added a basic infrastructure for writing helpers that are only compiled in the unit tests. Basically any C file you add to `test/unit/helpers` will be built into the test library, so you can use it to add wrappers for iinline functions

oni-link commented:
Using clang-format on the whole file eval.c (`:'<,'>pyf ~/git/llvm/tools/clang/tools/clang-format/clang-format.py `) gives the following error:
```
Error detected while processing function provider#python#Call:
line   12:
Channel was closed by the client
```
No error on master.

oni-link commented on src/nvim/os/input.c:
```diff
@@ -302,16 +302,16 @@ static InbufPollResult inbuf_poll(int ms)
-  (void)rbuffer_write(input_buffer, buf, buf_size);
-  rbuffer_consumed(read_buffer, buf_size);
+  RBUFFER_UNTIL_READ_OR_EMPTY(read_buffer, ptr, len) {
+    (void)rbuffer_write(input_buffer, ptr, len);
```
If `input_buffer` is not read for a while and is (nearly) full, we could loose some bytes.

tarruda replied:
This bug is already present(#1813) right? I intend to fix before #2371, which will be done by allocating memory for new input.

oni-link commented on src/nvim/os/input.c:
```diff
@@ -320,20 +320,21 @@ static void process_interrupts(void)
 
-  char *inbuf = rbuffer_read_ptr(input_buffer);
-  size_t count = rbuffer_pending(input_buffer), consume_count = 0;
+  RBUFFER_UNTIL_READ_OR_EMPTY(input_buffer, inbuf, count) {
```
That does not work if the buffer wraps around and in the first half (of the read section) a `3` was found:
In the first loop iteration the first half is cleared up to the `3`. The next iteration searches in the remainder of the first half. But because we do at most two iterations the second half is not searched for a `3`.

oni-link commented on src/nvim/os/rstream_defs.h:
```diff
@@ -3,15 +3,18 @@
 /// Type of function called when the RStream receives data
 ///
 /// @param rstream The RStream instance
+/// @param rbuffer The associated RBuffer isntance instance
```
`isntance` can be removed.

tarruda replied:
:+1:

jszakmeister commented:
> @tarruda Nope. My rb_insert/rb_remove support inserting at/removing from any position.

I'm just curious @ZyX-I, but that doesn't sound like a typical ring buffer behavior--it seems more list-like.  Why do you want this in a ring buffer?  Isn't it inefficient too--O(n) instead of O(1)--when not at the "end" (or "beginning" for remove) of the buffer?  I'm sure you have your reasons, I'm just curious about the motivations.

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -281,25 +281,22 @@ static void dynamic_buffer_ensure(DynamicBuffer *buf, size_t desired)
-                                rbuffer_pending(rbuffer), false, eof);
-  rbuffer_consumed(rbuffer, written);
+  RBUFFER_UNTIL_READ_OR_EMPTY(buf, ptr, len) {
+    rbuffer_consumed(buf, write_output(ptr, len, false, eof));
```
Does not work for the wrap around case:
If in the first half no `NL` and `NUL` are found, `write_output()` returns `0`. So in the second (and last) iteration of the loop the first half is used again, but the second half is not used (possible `NL` here).
If a `NUL` was found in the first iteration, it would be changed to a `NL` and the next iteration would
output "something".

Is `write_output()` working correctly for `eof == true`, if the two buffer halfs are written not as one?

oni-link commented on src/nvim/tui/term_input.inl:
```diff
@@ -162,54 +162,56 @@ static void timer_cb(uv_timer_t *handle)
-        // Don't mess with the paste option
+  RBUFFER_UNTIL_READ_OR_EMPTY(input->read_buffer, ptr, len) {
+    if (len > 5 && (!strncmp(ptr, "\x1b[200~", 6)
+          || !strncmp(ptr, "\x1b[201~", 6))) {
```
Does not work for the wrap around case:
A suffix of the string `"\x1b[200~"` could be in the second half.

oni-link commented on src/nvim/tui/term_input.inl:
```diff
@@ -162,54 +162,56 @@ static void timer_cb(uv_timer_t *handle)
-    tk_getkeys(input, true);
-    return true;
+  RBUFFER_UNTIL_READ_OR_EMPTY(input->read_buffer, ptr, len) {
+    if (len > 1 && ptr[0] == ESC && ptr[1] == NUL) {
```
Does not work for the wrap around case:
The `NUL` could be in the second half.

ZyX-I commented:
@jszakmeister It is O(n), but these are simple `memmove`s. The reasoning is that I a) avoid a huge amount of `malloc()` calls and b) it should happen rather rarely. Another options considered:

- If I will use a double linked list I will have to do one additional `malloc()` per element.
- If I will use a simple array then I will have to `memmove` each time element is added at the end once buffer is filled.

Though one thing looks like being worth trying: double linked list based on a pre-allocated array. This is two `malloc()` calls per list (list itself + array of pointers to “free” list items) (can be refactored to one malloc() if absolutely needed) and `O(1)` for inserting/removing at any position. Did not came up with this idea until your question and ring buffer seemed more natural for history list.

ShaDa code needs iteration, insert/append, remove; NeoVim history needs iteration, append, remove, index (take element with given index). So for ShaDa linked list is completely fine (insert and remove positions are determined using iteration, so there is no difference between ring buffer and array here), for NeoVim it is O(n) index, but most of time nobody cares though: index is AFAIK for `histget(, -x)` only (`histget(, +x)` needs iteration). Maybe it is worth refactoring in both places.

oni-link commented on src/nvim/tui/term_input.inl:
```diff
@@ -236,19 +238,20 @@ static void read_cb(RStream *rstream, void *rstream_data, bool eof)
-      if (ptr[i] == '\x1b') {
-        break;
+
+    RBUFFER_UNTIL_READ_OR_EMPTY(input->read_buffer, ptr, len) {
```
Can't we use `rbuffer_read_ptr()` instead of the `RBUFFER_UNTIL_READ_OR_EMPTY` macro?
`handle_bracket_*()` is otherwise not checked after something was consumed.

justinmk commented:
> Basically any C file you add to test/unit/helpers will be built into the test library,

This may cause more issues with gcov, until https://github.com/neovim/neovim/issues/2632 is fixed.

tarruda commented:
> This may cause more issues with gcov, until #2632 is fixed.

Opened #2682 to fix it

tarruda commented:
> Using clang-format on the whole file eval.c (:'<,'>pyf ~/git/llvm/tools/clang/tools/clang-format/clang-format.py) gives the following error:

@oni-link I have refactored this branch, the bugs when wrapping around should be gone. Can you test again if the clang format bug persists?

tarruda commented:
Hmm, just need a reverse-iterator macro to use instead of `RBUFFER_EACH`

oni-link commented on src/nvim/lib/rbuffer.h:
```diff
@@ -0,0 +1,285 @@
+    if (buf->write_ptr == buf->end_ptr) {
+      buf->write_ptr = buf->start_ptr;
+    }
+  }
```
Line 145-156 could be replaced with
```c
  buf->write_ptr += count;
  if (buf->write_ptr >= buf->end_ptr) {
     // wrap around
      buf->write_ptr -= rbuffer_capacity(buf);
  }
```

tarruda replied:
:+1: nice simplification, thanks

oni-link commented on src/nvim/lib/rbuffer.h:
```diff
@@ -0,0 +1,285 @@
+    if (buf->read_ptr == buf->end_ptr) {
+      buf->read_ptr = buf->start_ptr;
+    }
+  }
```
Lines 196-207 could be replaced with
```c
  buf->read_ptr += count;
  if (buf->read_ptr >= buf->end_ptr) {
      buf->read_ptr -= rbuffer_capacity(buf);
  }

```

tarruda replied:
:+1:

oni-link commented on src/nvim/lib/rbuffer.h:
```diff
@@ -0,0 +1,285 @@
+//   consumes data, return the number of bytes actually read:
+//
+//       RBUFFER_EACH_READ_PTR(rbuf, ptr, cnt)
+//         rbuffer_consumed(rbuf, read_data(state, ptr, cnt));
```
Does not work if `read_data()` does not returns `cnt`. The next read uses a wrong `ptr`.

tarruda replied:
I forgot to update the usage of these macros which have different semantics. These ones are only useful iterate over the all the data and only consume later:

```c
size_t consumed = 0;
RBUFFER_EACH_READ_PTR(rbuf, ptr, cnt)
  if ((consumed += read_data(state, ptr, cnt)) < cnt) break;
rbuffer_consumed(rbuf, consumed);
```

oni-link commented on src/nvim/lib/rbuffer.h:
```diff
@@ -0,0 +1,285 @@
+  for (size_t rcnt = 0, _r = buf->size; _r != SIZE_MAX; _r = SIZE_MAX)         \
+    for (char *rptr = rbuffer_read_ptr(buf, &rcnt);                            \
+         _r;                                                                   \
+         _r -= rcnt, rptr = _r ? buf->start_ptr : NULL, rcnt = _r ? _r : 0)
```
Would it be problematic to use the macro with an additional (code) argument? Seems to be cleaner, easier to check and does not have the problem with reads that consume less than `rcnt` bytes in one iteration (from the example above `RBUFFER_EACH_READ_PTR(rbuf, ptr, cnt) rbuffer_consumed(rbuf, read_data(state, ptr, cnt));`).
```c
#define RBUFFER_EACH_READ_PTR(buf, rptr, rcnt, code)                          \
  do {                                                                         \
    size_t rcnt;                                                               \
    while (buf->size) {                                                        \
      char *rptr = rbuffer_read_ptr(buf, &rcnt);                               \
      code                                                                     \
    }                                                                          \
  } while (0)
```
Usage is similar to the old macro:
```c
static inline size_t rbuffer_read(RBuffer *buf, char *dst, size_t dst_size)
  FUNC_ATTR_NONNULL_ALL
{
  size_t size = dst_size;
  RBUFFER_EACH_READ_PTR(buf, rptr, rcnt, {
    size_t copy_count = MIN(dst_size, rcnt);
    memcpy(dst, rptr, copy_count);
    rbuffer_consumed(buf, copy_count);

    if (!(dst_size -= copy_count)) {
      return size;
    }

    dst += copy_count;
  });

  return size - dst_size;
}
```

tarruda replied:
What you suggested has the same semantics as the previous `RBUFFER_UNTIL_EMPTY` macro. (See the response to the above comment for an updated usage).

oni-link commented:
> @oni-link I have refactored this branch, the bugs when wrapping around should be gone. Can you test again if the clang format bug persists?

The bug is still there.

ianks commented on src/nvim/lib/rbuffer.h:
```diff
@@ -0,0 +1,285 @@
+         _r;                                                                   \
+         _r -= wcnt, wptr = _r ? buf->start_ptr : NULL, wcnt = _r ? _r : 0)
+
+
```
Out of curiosity, what's the benefit of using a macro in these cases?

tarruda replied:
> Out of curiosity, what's the benefit of using a macro in these cases?

The alternative would be to use callback functions. But since C does not support closures, each time a ring buffer needed to be iterated a new function would need to be defined. IMO using a macro that can be followed by a block of code makes the result much cleaner and idiomatic.

oni-link commented:
>Using clang-format on the whole file eval.c (:'<,'>pyf ~/git/llvm/tools/clang/tools/clang-format/clang-format.py) gives the following error:
Error detected while processing function provider#python#Call:
line   12:
Channel was closed by the client
No error on master.


@tarruda, libuv does not always uses the allocated buffer for reading, http://docs.libuv.org/en/latest/stream.html#c.uv_read_start. The following patch fixes my problem:
```diff
diff --git a/src/nvim/os/rstream.c b/src/nvim/os/rstream.c
index f5f4754..1b5bb44 100644
--- a/src/nvim/os/rstream.c
+++ b/src/nvim/os/rstream.c
@@ -179,8 +179,9 @@ static void read_cb(uv_stream_t *stream, ssize_t cnt, const uv_buf_t *buf)
   RStream *rstream = handle_get_rstream((uv_handle_t *)stream);
 
   if (cnt <= 0) {
-    if (cnt != UV_ENOBUFS) {
-      DLOG("Closing RStream(%p)", rstream);
+    if (cnt != UV_ENOBUFS && cnt != 0) {
+      DLOG("Closing RStream(%p) because of %s(%zd)", rstream,
+           uv_strerror((int)cnt), cnt);
       // Read error or EOF, either way stop the stream and invoke the callback
       // with eof == true
       uv_read_stop(stream);
```

tarruda commented:
> @tarruda, libuv does not always uses the allocated buffer for reading, http://docs.libuv.org/en/latest/stream.html#c.uv_read_start. The following patch fixes my problem:

@oni-link not sure if I could have found this by myself, thank you for this great help. About the `uv_read_start` note:

> Note: nread might also be 0, which does not indicate an error or EOF, it happens when libuv requested a buffer through the alloc callback but then decided that it didn’t need that buffer.

Can you see why this bug only appeared now?

oni-link commented:
@tarruda, in the old code libuv never needed to allocate a buffer that then was not used. `alloc_cb()` always returned a buffer that was so big, that a libuv read could not fill it before `read_cb()` was called. So no need to allocate a new buffer. 

In the new code `alloc_cb()` returns smaller and smaller buffers until the ring buffer wraps around (until then the free space is split into two parts). So when the free space at the end of the buffer is completely filled with **one** read, an unused buffer will be allocated.

I could reproduce this "new" libuv behaviour with the following patch on master:
```diff
diff --git a/src/nvim/os/rstream.c b/src/nvim/os/rstream.c
index 702f282..cb91590 100644
--- a/src/nvim/os/rstream.c
+++ b/src/nvim/os/rstream.c
@@ -326,7 +326,7 @@ static void alloc_cb(uv_handle_t *handle, size_t suggested, uv_buf_t *buf)
 {
   RStream *rstream = handle_get_rstream(handle);
 
-  buf->len = rbuffer_available(rstream->buffer);
+  buf->len = MIN(31,rbuffer_available(rstream->buffer));
   buf->base = rbuffer_write_ptr(rstream->buffer);
 }

```

oni-link commented on src/nvim/rbuffer.c:
```diff
@@ -0,0 +1,201 @@
+    return rv;
+  }
+
+  return memcmp(str + n, buf->start_ptr, MIN(count, remaining));
```
If always `count <= buf->size`, `MIN(...)` could be replaced with `count`. Otherwise a longer string could still match a shorter buffer. Is this intentional (see unittest)?

splinterofchaos replied:
If the string is longer than the buffer, how do we know the rest of it isn't coming down the pipe?

tarruda replied:
> If always count <= buf->size, MIN(...) could be replaced with count. Otherwise a longer string could still match a shorter buffer. Is this intentional (see unittest

:+1: , I will just add an assertion for count <= buf->size(the caller should check `rbuffer_size`)

> If the string is longer than the buffer, how do we know the rest of it isn't coming down the pipe?

This is only supposed to compare the current RBuffer contents

tarruda commented:
> @tarruda, in the old code libuv never needed to allocate a buffer that then was not used. alloc_cb() always returned a buffer that was so big, that a libuv read could not fill it before read_cb() was called. So no need to allocate a new buffer.

> In the new code alloc_cb() returns smaller and smaller buffers until the ring buffer wraps around (until then the free space is split into two parts). So when the free space at the end of the buffer is completely filled with one read, an unused buffer will be allocated.

Makes perfect sense and explains why setting `JOB_BUFFER_SIZE`/`CHANNEL_BUFFER_SIZE` to 1 breaks without your patch.

oni-link commented on src/nvim/rbuffer.c:
```diff
@@ -0,0 +1,201 @@
+#endif
+
+/// Creates a new `RBuffer` instance.
+RBuffer *rbuffer_new(size_t capacity)
```
Function attributes: FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_NONNULL_RET

tarruda replied:
:+1:

oni-link commented on src/nvim/rbuffer.c:
```diff
@@ -0,0 +1,201 @@
+  size_t rcnt;
+  (void)rbuffer_read_ptr(buf, &rcnt);
+  return index < rcnt ?  buf->read_ptr + index : buf->start_ptr + index - rcnt;
+}
```
A version that computes the pointer as in `rbuffer_consumed()`:
```diff
 diff --git a/src/nvim/rbuffer.c b/src/nvim/rbuffer.c
index 4069703..e1ecc7e 100644
--- a/src/nvim/rbuffer.c
+++ b/src/nvim/rbuffer.c
@@ -175,12 +175,15 @@ size_t rbuffer_read(RBuffer *buf, char *dst, size_t dst_size)
   return size - dst_size;
 }
 
-char *rbuffer_get(RBuffer *buf, size_t index) FUNC_ATTR_NONNULL_ALL
+char *rbuffer_get(RBuffer *buf, size_t index)
+    FUNC_ATTR_NONNULL_ALL FUNC_ATTR_NONNULL_RET
 {
   assert(index < buf->size);
-  size_t rcnt;
-  (void)rbuffer_read_ptr(buf, &rcnt);
-  return index < rcnt ?  buf->read_ptr + index : buf->start_ptr + index - rcnt;
+  char *rptr = buf->read_ptr + index;
+  if (rptr >= buf->end_ptr) {
+    rptr -= rbuffer_capacity(buf);
+  }
+  return rptr;
 }
 
 int rbuffer_cmp(RBuffer *buf, const char *str, size_t count) FUNC_ATTR_NONNULL_ALL
```

tarruda replied:
:+1:

tarruda committed:
test: Don't run legacy test 87

The python3 emulation layer doesn't work well enough to run that test. Also add
notes to test86/test87 explaining why.

tarruda committed:
os/fs: Use module-local uv_loop_t instance

This event loop is just a stub instance used in synchronous libuv function
calls, it needs to be decoupled from the main event loop in order to run it from
another thread.

tarruda committed:
klib: Improve klist.h

- Add `kl_shift_at` macro and backing function. This can be used to shift
  elements at arbitrary positions. `kl_shift` is now defined on top of the new
  macro.
- Change shift/push API, now `kl_push` accepts an object as parameter and
  `kl_shift` returns the object instead of a status. An assertion against
  shifting at the end of a list(or empty lists) was added.
- Add `kl_iter` and `kl_iter_at` macros. `kl_iter_at` is for starting the
  iteration at arbitrary positions.

oni-link committed:
rstream: Fix bug triggered when libuv doesn't use the allocated buffer

Libuv will return 0 to signal that the buffer allocated by `alloc_cb` wasn't
used, and in this case the read_cb should simply be ignored.

oni-link commented on test/unit/rbuffer_spec.lua:
```diff
@@ -0,0 +1,354 @@
+        eq(0, cmp('long'))
+        eq(-1, cmp('lonG'))
+        eq(1, cmp('lonh'))
+        eq(0, cmp('long string2'))
```
This test will now fail, because the string is longer than the buffer content.

tarruda replied:
:+1:

oni-link commented on test/unit/rbuffer_spec.lua:
```diff
@@ -0,0 +1,354 @@
+
+    describe('with empty buffer', function()
+      it('returns 0 since no characters are compared', function()
+        eq(0, cmp('string'))
```
This test will now fail, because the string is longer than the buffer content.

tarruda replied:
:+1:

oni-link commented on src/nvim/rbuffer.h:
```diff
@@ -0,0 +1,77 @@
+// - Pass the write pointer to a function(`write_data`) that incrementally
+//   produces data, returning the number of bytes actually written:
+//
+//       RBUFFER_EACH_WRITE_PTR(rbuf, ptr, cnt)
```
Macro name need updating.

tarruda replied:
:+1:

oni-link commented on src/nvim/rbuffer.h:
```diff
@@ -0,0 +1,77 @@
+// - Pass the read pointer to a function(`read_data`) that incrementally
+//   consumes data, return the number of bytes actually read:
+//
+//       RBUFFER_EACH_READ_PTR(rbuf, ptr, cnt)
```
Macro name need updating.

tarruda replied:
:+1:

oni-link commented on src/nvim/tui/term_input.inl:
```diff
@@ -236,19 +235,25 @@ static void read_cb(RStream *rstream, void *rstream_data, bool eof)
-    for (i = ptr[0] == ESC ? 1 : 0; i < len; i++) {
-      if (ptr[i] == '\x1b') {
+    size_t count;
+    RBUFFER_EACH(buf, c, i) {
```
`input->read_buffer` instead of `buf`?

The new code pushes everthing up to the next 'esc' (including).

tarruda replied:
:+1:

oni-link commented on src/nvim/tui/term_input.inl:
```diff
@@ -236,19 +235,25 @@ static void read_cb(RStream *rstream, void *rstream_data, bool eof)
+    RBUFFER_UNTIL_EMPTY(input->read_buffer, ptr, len) {
+      size_t consumed = termkey_push_bytes(input->tk, ptr, MIN(count, len));
+      rbuffer_consumed(input->read_buffer, consumed);
+      tk_getkeys(input, false);
```
This should be placed after the macro.

tarruda replied:
If I put `tk_getkeys` after, it's possible that libtermkey input buffer will be filled before `count` is consumed, so we'll enter an infinite loop(`termkey_push_bytes` will return 0, waiting for `tk_getkeys` to process the buffered data)

tarruda replied:
>  so we'll enter an infinite loop

In a debug build it will just abort due to a failed assertion in `rbuffer_consumed`

oni-link replied:
I think `termkey_push_bytes()` can never return `0`. Either `(size_t) -1` or `len > 0` is returned.

justinmk commented on src/nvim/memory.h:
```diff
@@ -1,6 +1,7 @@
 #define NVIM_MEMORY_H
 
+#include <stdint.h>  // for size_t
 #include <stddef.h>  // for size_t
```
one of the `// for size_t` comments must be wrong

tarruda replied:
`yyp` error :smile:

oni-link commented on src/nvim/os/shell.c:
```diff
@@ -281,25 +281,27 @@ static void dynamic_buffer_ensure(DynamicBuffer *buf, size_t desired)
-                                rbuffer_pending(rbuffer), false, eof);
-  rbuffer_consumed(rbuffer, written);
+  RBUFFER_UNTIL_EMPTY(buf, ptr, len) {
+    size_t written = write_output(ptr, len, false, eof);
```
Can we call `write_output()` two times(wrap around case) with the `eof` flag set? Is the second buffer half corretly appended?

tarruda replied:
Fixed by checking that `len <= rbuffer_size(buf)`

justinmk commented on test/unit/helpers/rbuffer.h:
```diff
@@ -0,0 +1,9 @@
+#include "nvim/rbuffer.h"
```
I think the more descriptive name in testing parlance is "fixtures". In the functional tests we have `test/functional/fixtures/`, so this new path might as well be renamed to `test/unit/fixtures/`.

tarruda replied:
:+1:

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -40,6 +40,7 @@ file(MAKE_DIRECTORY ${GENERATED_INCLUDES_DIR}/tui)
 file(GLOB NEOVIM_SOURCES *.c os/*.c api/*.c api/private/*.c msgpack_rpc/*.c
   tui/*.c)
 file(GLOB_RECURSE NEOVIM_HEADERS *.h)
+file(GLOB UNIT_TEST_HELPERS ${PROJECT_SOURCE_DIR}/test/unit/helpers/*.c)
```
`UNIT_TEST_FIXTURES`

tarruda replied:
:+1:

oni-link commented:
Just noticed:
Going through the history with `UP` does not work anymore in command mode. Instead nvim changes from command mode to normal mode.

oni-link commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -328,19 +328,19 @@ static void channel_from_stdio(void)
 
-  while ((count = rstream_pending(rstream))) {
-    size_t read = rstream_read(rstream, buf, sizeof(buf) - 1);
+  while ((count = rbuffer_size(rbuf))) {
```
Variable `count` can be removed.

tarruda replied:
:+1:

oni-link commented on src/nvim/tui/term_input.inl:
```diff
@@ -236,19 +235,25 @@ static void read_cb(RStream *rstream, void *rstream_data, bool eof)
+    size_t count;
+    RBUFFER_EACH(buf, c, i) {
+      count = i + 1;
+      if (c == '\x1b') {
```
This restores the old push behaviour and fixes the problem with the history in command mode:
`if (c == '\x1b' && i > 0) {`

tarruda replied:
thanks

oni-link commented on src/nvim/tui/term_input.inl:
```diff
@@ -236,19 +235,29 @@ static void read_cb(RStream *rstream, void *rstream_data, bool eof)
+    }
+
+    RBUFFER_UNTIL_EMPTY(input->read_buffer, ptr, len) {
+      size_t consumed = termkey_push_bytes(input->tk, ptr, MIN(count, len));
```
`termkey_push_bytes()` could return a value bigger than `input->read_buffer->size`, so the following call to `rbuffer_consumed()` would fail.

tarruda replied:
How can that happen? We are passing `MIN(count, len)` to `termkey_push_bytes` which should always be <= `input->read_buffer->size`

oni-link replied:
If the termkey buffer gets full it will return (size_t)-1.
```c
size_t termkey_push_bytes(TermKey *tk, const char *bytes, size_t len)
{
  if(tk->buffstart) {
    memmove(tk->buffer, tk->buffer + tk->buffstart, tk->buffcount);
    tk->buffstart = 0;
  }

  /* Not expecting it ever to be greater but doesn't hurt to handle that */
  if(tk->buffcount >= tk->buffsize) {
    errno = ENOMEM;
    return (size_t)-1;
  }

  if(len > tk->buffsize - tk->buffcount)
    len = tk->buffsize - tk->buffcount;

  // memcpy(), not strncpy() in case of null bytes in input
  memcpy(tk->buffer + tk->buffcount, bytes, len);
  tk->buffcount += len;

  return len;
}

```

tarruda replied:
I'm not sure on how to handle this error, my  best guess is to treat as an out of memory condition and exit the program. What do you think?

@leonerd, is this `tk->buffcount >= tk->buffsize` condition related to libtermkey failing to allocate memory at some point?

leonerd replied:
Not so much failing, as simply trying to keep a bounded size. I felt it best not to grow the buffer arbitrarily.

tarruda replied:
> Not so much failing, as simply trying to keep a bounded size. I felt it best not to grow the buffer arbitrarily.

ok, but if the caller consumes all parsed keys right after every `termkey_push_bytes` call, then this condition is not possible right?

tarruda committed:
rbuffer: Reimplement as a ring buffer and decouple from rstream

Extract the RBuffer class from rstream.c and reimplement it as a ring buffer,
a more efficient version that doesn't need to relocate memory.

The old rbuffer_read/rbuffer_write interfaces are kept for simple
reading/writing, and the RBUFFER_UNTIL_{FULL,EMPTY} macros are introduced to
hide wrapping logic when more control is required(such as passing the buffer
pointer to a library function that writes directly to the pointer)

Also add a basic infrastructure for writing helper C files that are only
compiled in the unit test library, and use this to write unit tests for RBuffer
which contains some macros that can't be accessed directly by luajit.

Helped-by: oni-link <knil.ino@gmail.com>
Reviewed-by: oni-link <knil.ino@gmail.com>
Reviewed-by: Scott Prager <splinterofchaos@gmail.com>
Reviewed-by: Justin M. Keyes <justinkz@gmail.com>
Reviewed-by: Michael Reed <m.reed@mykolab.com>
